### PR TITLE
NonlinearRZ ref #4706

### PR DESCRIPTION
**Pull request by jasondhales**
Add option for full nonlinear RZ (including incremental rotation).

### CODE REVIEW

File: modules/solid_mechanics/src/materials/Nonlinear.C [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+ ////   const Real log2 = std::log(lamda2);
+ ////   const Real log3 = std::log(lamda3);
+ ////
+ ////   _Uhat = N1 * N1.transpose() * lamda1 +  N2 * N2.transpose() * lamda2 +  N3 * N3.transpose() * lamda3;
+ ////
+ ////   invertMatrix(_Uhat,invUhat);
+ ////
+ ////   _incremental_rotation = Fhat * invUhat;
+ ////
+ ////   strain_increment = N1 * N1.transpose() * log1 +  N2 * N2.transpose() * log2 +  N3 * N3.transpose() * log3;
```
Comment (jwpeterson) [inline modules/solid_mechanics/src/materials/Nonlinear.C]: Do you need to keep the preceding 30 commented out lines?  Looks like an initial implementation that got moved elsewhere?

File: modules/solid_mechanics/src/materials/Nonlinear.C [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+
+}
+
+////////////////////////////////////////////////////////////////////////
+
+void
+Nonlinear::finalizeStress( std::vector<SymmTensor*> & t)
+{
+  // Using the incremental rotation, update the stress to the current configuration (R*T*R^T)
+  for (unsigned i(0); i < t.size(); ++i)
```
Comment (jwpeterson) [inline modules/solid_mechanics/src/materials/Nonlinear.C]: Integer initialization coding convention.

File: modules/combined/tests/elastic_patch/elastic_patch_rz_nonlinear.i [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+# The code computes stress = 19900.6617 when
+# increment_calculation = eigen.  There is a small error when the
+# rashidapprox option is used.
+#
+
+[Problem]
+  coord_type = RZ
+[]
+
+[Mesh]#Comment
```
Comment (jwpeterson) [inline modules/combined/tests/elastic_patch/elastic_patch_rz_nonlinear.i]: Remove "#Comment"

**Comment by jwpeterson** [comment]
Looks good other than the few things mentioned.  A request for the future would be to split code and tests into separate commits; it makes reviewing the diffs a bit easier.

**Comment by friedmud** [comment]
A bit tangential... but @jasondhales how difficult would it be solid_mechanics to support the new capability to do RZ about any arbitrary axis that we added here: https://github.com/idaholab/moose/pull/4533  ?

I actually have a problem where I would like to use that capability...

**Comment by jasondhales** [comment]
At least these three would need adjusting: AxisymmetricRZ, NonlinearRZ, StressDivergenceRZ.  Also, a default has been chosen in InternalVolume that makes it work properly for standard RZ.  Just curious, what's the motivation for setting the axis instead of just using y?  The Issue didn't give any information.

**Comment by friedmud** [comment]
It matches up with the physical configuration we have (ie... we have a pipe laying on it's side in the real world).

**Pull request merged**